File: TOMIMSV4/DAL/Transacciones/Stock_Parametro/clsLnStock_parametro.vb
Imports System.Data.SqlClient
Imports System.Reflection

Public Class clsLnStock_parametro

    Public Shared Sub Cargar(ByRef oBeStock_parametro As clsBeStock_parametro, ByRef dr As DataRow)
        Try
            With oBeStock_parametro
                .IdStockParametro = IIf(IsDBNull(dr.Item("IdStockParametro")), 0, dr.Item("IdStockParametro"))
                .IdStock = IIf(IsDBNull(dr.Item("IdStock")), 0, dr.Item("IdStock"))
                .IdProductoParametro = IIf(IsDBNull(dr.Item("IdProductoParametro")), 0, dr.Item("IdProductoParametro"))
                .Valor_texto = IIf(IsDBNull(dr.Item("valor_texto")), "", dr.Item("valor_texto"))
                .Valor_numerico = IIf(IsDBNull(dr.Item("valor_numerico")), 0.0, dr.Item("valor_numerico"))
                .Valor_fecha = IIf(IsDBNull(dr.Item("valor_fecha")), Date.Now, dr.Item("valor_fecha"))
                .Valor_logico = IIf(IsDBNull(dr.Item("valor_logico")), False, dr.Item("valor_logico"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
            End With

        Catch ex1 As SQLException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try
    End Sub

    Public Shared Function Insertar(ByRef oBeStock_parametro As clsBeStock_parametro, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Ins.Init("stock_parametro")
            Ins.Add("idstockparametro", "@idstockparametro", DataType.Parametro)
            Ins.Add("idstock", "@idstock", DataType.Parametro)
            Ins.Add("idproductoparametro", "@idproductoparametro", DataType.Parametro)
            Ins.Add("valor_texto", "@valor_texto", DataType.Parametro)
            Ins.Add("valor_numerico", "@valor_numerico", DataType.Parametro)
            Ins.Add("valor_fecha", "@valor_fecha", DataType.Parametro)
            Ins.Add("valor_logico", "@valor_logico", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)

            Dim sp As String = Ins.SQL()
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDSTOCKPARAMETRO", oBeStock_parametro.IdStockParametro))
            cmd.Parameters.Add(New SqlParameter("@IDSTOCK", oBeStock_parametro.IdStock))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETRO", oBeStock_parametro.IdProductoParametro))
            cmd.Parameters.Add(New SqlParameter("@VALOR_TEXTO", oBeStock_parametro.Valor_texto))
            cmd.Parameters.Add(New SqlParameter("@VALOR_NUMERICO", oBeStock_parametro.Valor_numerico))
            cmd.Parameters.Add(New SqlParameter("@VALOR_FECHA", oBeStock_parametro.Valor_fecha))
            cmd.Parameters.Add(New SqlParameter("@VALOR_LOGICO", oBeStock_parametro.Valor_logico))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeStock_parametro.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeStock_parametro.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeStock_parametro.Activo))


            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeStock_parametro As clsBeStock_parametro, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Upd.Init("stock_parametro")
            Upd.Add("idstockparametro", "@idstockparametro", DataType.Parametro)
            Upd.Add("idstock", "@idstock", DataType.Parametro)
            Upd.Add("idproductoparametro", "@idproductoparametro", DataType.Parametro)
            Upd.Add("valor_texto", "@valor_texto", DataType.Parametro)
            Upd.Add("valor_numerico", "@valor_numerico", DataType.Parametro)
            Upd.Add("valor_fecha", "@valor_fecha", DataType.Parametro)
            Upd.Add("valor_logico", "@valor_logico", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Where("IdStockParametro = @IdStockParametro")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDSTOCKPARAMETRO", oBeStock_parametro.IdStockParametro))
            cmd.Parameters.Add(New SqlParameter("@IDSTOCK", oBeStock_parametro.IdStock))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETRO", oBeStock_parametro.IdProductoParametro))
            cmd.Parameters.Add(New SqlParameter("@VALOR_TEXTO", oBeStock_parametro.Valor_texto))
            cmd.Parameters.Add(New SqlParameter("@VALOR_NUMERICO", oBeStock_parametro.Valor_numerico))
            cmd.Parameters.Add(New SqlParameter("@VALOR_FECHA", oBeStock_parametro.Valor_fecha))
            cmd.Parameters.Add(New SqlParameter("@VALOR_LOGICO", oBeStock_parametro.Valor_logico))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeStock_parametro.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeStock_parametro.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeStock_parametro.Activo))


            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
        End Try

    End Function


    Public Shared Function Eliminar(ByRef oBeStock_parametro As clsBeStock_parametro, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Stock_parametro" &
             "  Where(IdStockParametro = @IdStockParametro)"

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDSTOCKPARAMETRO", oBeStock_parametro.IdStockParametro))


            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
        End Try

    End Function

    Public Shared Function EliminarTodos(Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Stock_parametro"
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
        End Try

    End Function

    Public Shared Function MaxID() As Integer

        Try

            Dim lMax As Integer = 0

            Const sp As String = "SELECT ISNULL(Max(IdStockParametro),0) FROM Stock_parametro"

            Using lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)

                    Using lCommand As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}

                        Dim lReturnValue As Object = lCommand.ExecuteScalar()

                        If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                            lMax = CInt(lReturnValue)
                        End If

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

            Return lMax

        Catch ex1 As SQLException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

End Class

File: WMS.DALCore/Stock/clsLnStock_parametro.cs
namespace WMS.DALCore.Stock
{
    using AppGlobal;
    using Microsoft.Data.SqlClient;
    using Microsoft.Extensions.Configuration;
    using System;
    using System.Data;  
    using WMS.EntityCore.Stock;

    public class clsLnStock_parametro
    {
        private static clsInsert Ins = new clsInsert();
        private static clsUpdate Upd = new clsUpdate();
        public static void Cargar(ref clsBeStock_parametro oBeStock_parametro, DataRow dr)
        {
            if (oBeStock_parametro == null)
                throw new ArgumentNullException(nameof(oBeStock_parametro));

            if (dr == null)
                throw new ArgumentNullException(nameof(dr));

            try
            {
                // Función local para manejar valores de forma segura
                T GetSafeValue<T>(string columnName, T defaultValue, Func<object, T> converter)
                {
                    if (dr.Table.Columns.Contains(columnName) && !Convert.IsDBNull(dr[columnName]))
                    {
                        var value = dr[columnName];
                        if (value != null)
                            return converter(value);
                    }
                    return defaultValue;
                }

                oBeStock_parametro.IdStockParametro = GetSafeValue("IdStockParametro", 0, Convert.ToInt32);
                oBeStock_parametro.IdStock = GetSafeValue("IdStock", 0, Convert.ToInt32);
                oBeStock_parametro.IdProductoParametro = GetSafeValue("IdProductoParametro", 0, Convert.ToInt32);
                oBeStock_parametro.Valor_texto = GetSafeValue("valor_texto", "", Convert.ToString) ?? "";
                oBeStock_parametro.Valor_numerico = GetSafeValue("valor_numerico", 0.0, Convert.ToDouble);
                oBeStock_parametro.Valor_fecha = GetSafeValue("valor_fecha", DateTime.Now, Convert.ToDateTime);
                oBeStock_parametro.Valor_logico = GetSafeValue("valor_logico", false, Convert.ToBoolean);
                oBeStock_parametro.User_agr = GetSafeValue("user_agr", "", Convert.ToString) ?? "";
                oBeStock_parametro.Fec_agr = GetSafeValue("fec_agr", DateTime.Now, Convert.ToDateTime);
                oBeStock_parametro.Activo = GetSafeValue("activo", false, Convert.ToBoolean);
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static int Insertar(clsBeStock_parametro oBeStock_parametro, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (oBeStock_parametro == null)
                throw new ArgumentNullException(nameof(oBeStock_parametro));

            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                Ins.Init("stock_parametro");
                Ins.Add("idstockparametro", "@idstockparametro", "F");
                Ins.Add("idstock", "@idstock", "F");
                Ins.Add("idproductoparametro", "@idproductoparametro", "F");
                Ins.Add("valor_texto", "@valor_texto", "F");
                Ins.Add("valor_numerico", "@valor_numerico", "F");
                Ins.Add("valor_fecha", "@valor_fecha", "F");
                Ins.Add("valor_logico", "@valor_logico", "F");
                Ins.Add("user_agr", "@user_agr", "F");
                Ins.Add("fec_agr", "@fec_agr", "F");
                Ins.Add("activo", "@activo", "F");

                string sp = Ins.SQL();

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;

                    cmd.Parameters.Add(new SqlParameter("@IDSTOCKPARAMETRO", oBeStock_parametro.IdStockParametro));
                    cmd.Parameters.Add(new SqlParameter("@IDSTOCK", oBeStock_parametro.IdStock));
                    cmd.Parameters.Add(new SqlParameter("@IDPRODUCTOPARAMETRO", oBeStock_parametro.IdProductoParametro));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_TEXTO", oBeStock_parametro.Valor_texto));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_NUMERICO", oBeStock_parametro.Valor_numerico));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_FECHA", oBeStock_parametro.Valor_fecha));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_LOGICO", oBeStock_parametro.Valor_logico));
                    cmd.Parameters.Add(new SqlParameter("@USER_AGR", oBeStock_parametro.User_agr));
                    cmd.Parameters.Add(new SqlParameter("@FEC_AGR", oBeStock_parametro.Fec_agr));
                    cmd.Parameters.Add(new SqlParameter("@ACTIVO", oBeStock_parametro.Activo));

                    return cmd.ExecuteNonQuery();
                }
            }
            catch (SqlException)
            {
                throw;
            }
        }

        public static int Actualizar(clsBeStock_parametro oBeStock_parametro, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (oBeStock_parametro == null)
                throw new ArgumentNullException(nameof(oBeStock_parametro));

            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                Upd.Init("stock_parametro");
                Upd.Add("idstockparametro", "@idstockparametro", "F");
                Upd.Add("idstock", "@idstock", "F");
                Upd.Add("idproductoparametro", "@idproductoparametro", "F");
                Upd.Add("valor_texto", "@valor_texto", "F");
                Upd.Add("valor_numerico", "@valor_numerico", "F");
                Upd.Add("valor_fecha", "@valor_fecha", "F");
                Upd.Add("valor_logico", "@valor_logico", "F");
                Upd.Add("user_agr", "@user_agr", "F");
                Upd.Add("fec_agr", "@fec_agr", "F");
                Upd.Add("activo", "@activo", "F");
                Upd.Where("IdStockParametro = @IdStockParametro");

                string sp = Upd.SQL();

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;

                    cmd.Parameters.Add(new SqlParameter("@IDSTOCKPARAMETRO", oBeStock_parametro.IdStockParametro));
                    cmd.Parameters.Add(new SqlParameter("@IDSTOCK", oBeStock_parametro.IdStock));
                    cmd.Parameters.Add(new SqlParameter("@IDPRODUCTOPARAMETRO", oBeStock_parametro.IdProductoParametro));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_TEXTO", oBeStock_parametro.Valor_texto));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_NUMERICO", oBeStock_parametro.Valor_numerico));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_FECHA", oBeStock_parametro.Valor_fecha));
                    cmd.Parameters.Add(new SqlParameter("@VALOR_LOGICO", oBeStock_parametro.Valor_logico));
                    cmd.Parameters.Add(new SqlParameter("@USER_AGR", oBeStock_parametro.User_agr));
                    cmd.Parameters.Add(new SqlParameter("@FEC_AGR", oBeStock_parametro.Fec_agr));
                    cmd.Parameters.Add(new SqlParameter("@ACTIVO", oBeStock_parametro.Activo));

                    return cmd.ExecuteNonQuery();
                }
            }
            catch (SqlException)
            {
                throw;
            }
        }

        public static int Eliminar(clsBeStock_parametro oBeStock_parametro, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (oBeStock_parametro == null)
                throw new ArgumentNullException(nameof(oBeStock_parametro));

            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                const string sp = @"DELETE FROM Stock_parametro 
                               WHERE IdStockParametro = @IdStockParametro";

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;
                    cmd.Parameters.Add(new SqlParameter("@IDSTOCKPARAMETRO", oBeStock_parametro.IdStockParametro));
                    return cmd.ExecuteNonQuery();
                }
            }
            catch (SqlException)
            {
                throw;
            }
        }

        public static int EliminarTodos(SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                const string sp = "DELETE FROM Stock_parametro";

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;
                    return cmd.ExecuteNonQuery();
                }
            }
            catch (SqlException)
            {
                throw;
            }
        }

        public static int MaxID(IConfiguration config)
        {
            try
            {
                const string sp = "SELECT ISNULL(MAX(IdStockParametro), 0) FROM Stock_parametro";

                using (var lConnection = new SqlConnection(config.GetConnectionString("CST")))
                {
                    lConnection.Open();

                    using (var lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted))
                    {
                        using (var lCommand = new SqlCommand(sp, lConnection, lTransaction))
                        {
                            lCommand.CommandType = CommandType.Text;
                            object lReturnValue = lCommand.ExecuteScalar();

                            if (lReturnValue != DBNull.Value && lReturnValue != null)
                            {
                                return Convert.ToInt32(lReturnValue);
                            }
                            return 0;
                        }
                    }
                }
            }            
            catch (Exception)
            {                
                throw;
            }
        }
        public static int MaxID(SqlConnection lConnection, SqlTransaction lTransaction)
        {
            if (lConnection == null)
                throw new ArgumentNullException(nameof(lConnection));

            if (lTransaction == null)
                throw new ArgumentNullException(nameof(lTransaction));

            try
            {
                const string sp = "SELECT ISNULL(MAX(IdStockParametro), 0) FROM Stock_parametro";

                using (var lCommand = new SqlCommand(sp, lConnection, lTransaction))
                {
                    lCommand.CommandType = CommandType.Text;
                    object lReturnValue = lCommand.ExecuteScalar();

                    if (lReturnValue != DBNull.Value && lReturnValue != null)
                    {
                        return Convert.ToInt32(lReturnValue);
                    }
                    return 0;
                }
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static int Insertar_Stock_Parametro_Recepcion(clsBeStock_rec bo,
                                                             int IdStock,
                                                             SqlConnection lConnection,
                                                             SqlTransaction lTransaction)
        {
            if (bo == null)
                throw new ArgumentNullException(nameof(bo));

            if (lConnection == null)
                throw new ArgumentNullException(nameof(lConnection));

            if (lTransaction == null)
                throw new ArgumentNullException(nameof(lTransaction));

            int vRegistros = 0;

            try
            {
                int lMaxSP = MaxID(lConnection, lTransaction);

                var listadetp = clsLnTrans_re_det_parametros.Get_All_By_IdRecepcionEnc_And_IdRecepcionDet(bo.IdRecepcionEnc,
                                                                                                          bo.IdRecepcionDet,
                                                                                                          lConnection,
                                                                                                          lTransaction);

                if (listadetp != null && listadetp.Count > 0)
                {
                    foreach (var dp in listadetp)
                    {
                        if (dp == null)
                            continue;

                        var BeTransReDetParam = new clsBeStock_parametro();
                        clsPublic.CopyObject(dp, ref BeTransReDetParam);
                        lMaxSP += 1;
                        BeTransReDetParam.IdStockParametro = lMaxSP;
                        BeTransReDetParam.IdStock = IdStock;
                        vRegistros += Insertar(BeTransReDetParam, lConnection, lTransaction);
                    }
                }

                return vRegistros;
            }
            catch (Exception)
            {                
                throw;
            }
        }
    }
}
